Feature/6 replace unordered lists with table

aggiewb opened the pull request.

aggiewb committed:
Merge pull request #12 from aggiewb/feature/5-replace_object_literals_with_constructor_functions

Feature/5 replace object literals with constructor functions

BatemanVO committed:
Merge pull request #13 from aggiewb/release/2020-06-22

Release/2020 06 22

aggiewb committed:
Add table html elements

aggiewb committed:
Add function to createHoursRow() and update

location.simulatedCookiesPerHour property assignment to occur in calculateSimulatedCookiesPerHour()

aggiewb committed:
Remove functions to create HTML elements, and add function

createHoursRow()
Add css for table border and padding

aggiewb committed:
Add function createStoreRow() which add a row with a store

name and its simulated data to the table
Add function createStoreTableContents() which calls functions that create the content of the table
Update variable names to have more clarity, and add conditional to else if

aggiewb committed:
Update html h1 content, and add tr css

aggiewb committed:
Add function createTotalsRow() which adds a row of each

hour's sold donuts for all location
Call function in createStoreTableContents

aggiewb committed:
Change h1 content, and add js to update textContent of

caption table element

aggiewb committed:
Add function createStoreTotals() which add a column of daily

totals for each store
Add function createDailyLocationTotalHeader() which creates the header for the createStoreTotals() column
Add returns for functions to be able to use the row to add table data
Add function totalOfAllStoresDaily() to add all of the sold cookies for all stores on one day

aggiewb committed:
Change createStoreRow() to the render() method on the

CookieStore prototype as required in the ticket
Change all references of location in render() to this

aggiewb committed:
Add css for text and table color

aggiewb committed:
Add content to empty th in the thead

aggiewb committed:
Add the appropriate scope attributes to col or rows

aggiewb committed:
Change absolute sizes to proportional sizes

BatemanVO commented on css/style.css:
```diff
@@ -1,3 +1,20 @@
+    padding: 1.875rem 0.9375rem; /* 30px 15px */
+}
+
+tr:nth-child(even), tfoot{
```
What if we add one more store? The `tfoot` and the last row in the `tbody` would be the same color, which is not the appearance we want.

We may need to create a class that has this style and dynamically assign it to the `tfoot` depending on the number of stores added to the `tbody` - but we'll need some way of figuring out if we want to apply this background color to the `tfoot`, because half the time, we won't want to.

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
         var simulatedCookiesPerHour = [];
         for(var i = 0; i < openHours; i++){
-            simulatedCookiesPerHour[i] = Math.floor(location.randomCustomersPerHour() * location.averageCookiesPerCustomer);
+            location.simulatedCookiesPerHour[i] = Math.floor(location.randomCustomersPerHour() * location.averageCookiesPerCustomer);
```
I think this method should be refactored to be on the `CookieStore.prototype`. We're no longer making use of the `return` value, nor are we making use of the `simulatedCookiesPerHour` Array declared on line 25 - all we're doing is setting the `simulatedCookiesPerHour` property of a location. We could do the same thing by just making this a prototype method with no parameters and changing this line to be `this.simulatedCookiesPerHour[i] =`. Alternatively, we could just do `this.simulatedCookiesPerHour.push(`.

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
-            createStoreHeader(location);
-            createHourItems(location, createHoursList());
+    function createStoreTableContents(){
+        document.querySelector('caption').textContent = 'Cookies Needed By Location Each Day';
```
The text for this element can be hardcoded. The requirements weren't clear, but what they meant was that the _entire_ `<caption>` element can be hardcoded into the HTML.

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
+        var hoursOpen = 14;
+        
+        var tableRow = document.createElement('tr');    
+        document.querySelector('thead').appendChild(tableRow);
```
It'll be more efficient if you don't append the `tableRow` to the `thead` element until you're all done generating the `tableRow` content. Once you append something to the DOM, every change will [trigger a reflow](https://developers.google.com/speed/docs/insights/browser-reflow), whereas if you build everything the element needs, treating it like [a DocumentFragment](https://developer.mozilla.org/en/docs/Web/API/DocumentFragment), then append it at the end, it will only trigger one reflow.

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
+        var tableRow = document.createElement('tr');    
+        document.querySelector('thead').appendChild(tableRow);
+        var locationHeader = document.createElement('th');
+        locationHeader.textContent = 'Store Locations';
```
Since the first cell will not be blank, we should add a `scope="col"` attribute to this cell, too.

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
                 //1pm-11pm
-                newHourItem.textContent = (militaryHour - standardTimeOffSet) + 'pm: ' + hourSoldCookies + ' cookies';
+                newHourHead.textContent = (militaryHour - standardTimeOffSet) + ':00pm';
+            }
```
We could make this code a little more DRY by looking for common elements in these conditional blocks - we always reference `newHourHead.textContent =`, for example, so perhaps instead of typing it 4 times, we could instead declare a variable to represent what the String to assign to `newHourHead.textContent` is, then do so after the conditional blocks. Additionally, we reference `:00am` and `:00pm` twice - we could cache those before the loop and reference them as needed, too.

The code will lose a very, very slight bit of efficiency by doing so, but the maintenance and readability improvement should be worth the tradeoff.

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
+        dailyLocationTotal.setAttribute('scope', 'col');
+        dailyLocationTotal.textContent = 'Daily Location Total';
+        tableRow.appendChild(dailyLocationTotal);
     }
```
Perhaps we could refactor this function a little to use for both the top-left cell and the top-right cell of the table? For both, we're creating a `th`, setting its scope to `col`, and appending it to the `tableRow`, so the only difference is the `textContent` - which we could make a parameter. We could also make this code a little more functional by having it return the `th` element it creates instead of mutating a `tableRow` passed in.

So then, on line 56, we could do something like, `tableRow.appendChild(createHeaderCell('Store Locations'))` and on line 35, we could do `hoursTableRow.appendChild(createHeaderCell('Daily Location Total'))`.

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
+        var tableStoreTotalData = document.createElement('td');
+        tableStoreTotalData.textContent = total;
+        tableRow.appendChild(tableStoreTotalData);
+        return total;
```
While it's tempting to have the logic that appends the total to a row and the logic that keeps a running total of the row in the same function, the problem is that we're creating [side effects](https://en.wikipedia.org/wiki/Side_effect_(computer_science)). This function:
- Creates a store total `<td>` element,
- Mutates a `<tr>` element passed in, AND
- Sums the total for a given location, returning that value

So it's not readily apparent to future developers what the purpose of this function is, because it does more than what it's named for.

We could refactor to have something like:
![Capture](https://user-images.githubusercontent.com/20998142/85915111-adde7b00-b7f9-11ea-8dbb-31cd1d496b95.PNG)

In that way, we have a function that generates totals for a location (and really, `calculateTotal` should be a prototype method!), creates and returns `<td>` cells, then a "main" function that uses the return values of these smaller, single-purpose methods.

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
+
+    function createTotalsFooterRow(){
+        var totalsRow = document.createElement('tr');
+        totalsRow.setAttribute('scope', 'row');
```
`scope="row"` actually needs to be applied to the single `<th>` element _within_ each row, not to the row itself. A `<th>` element is used to indicate the head of a column or the head of a row (which is what the scope attribute does).

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
+        document.querySelector('tfoot').appendChild(totalsRow);
+        totalsRow.appendChild(totalsHead);
+
+        for(var i = 0; i < cookieStoreLocations[0].simulatedCookiesPerHour.length; i++){
```
This outer `for` loop is confusing - perhaps we should make a variable scoped to our `IIFE` called something like `openHours` that has a value of 14 or something like that. I looked at this for several minutes thinking it must be a bug, wondering why we were only referencing `cookieStoreLocations[0]`.

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
+        var totalOfAllStoresDailyTableData = document.createElement('td');
+        totalOfAllStoresDailyTableData.textContent = allStoresTotal;
+        tableRow.appendChild(totalOfAllStoresDailyTableData);
+    }
```
I think we should revisit all these functions that mutate a `<tr>` element and have them return a value that is then appended to the element, instead. It would remove a parameter and create code that relies on `return` values (referred to as [functional programming](https://en.wikipedia.org/wiki/Functional_programming)). Functional programming is generally easier to read, maintain, reuse, and scale.

BatemanVO commented on js/sales.js:
```diff
@@ -24,56 +24,119 @@
+        var tableStoreRow = document.createElement('tr');
+        tableStoreRow.setAttribute('scope', 'row');
+        var tableStoreHead = document.createElement('th');
+        tableStoreHead.textContent = this.name;
```
Good job setting the first non-empty cell of the row to a `<th>` element! We should move the `scope="row"` assignment to this `<th>` instead of on the `<tr>`, but I _constantly_ see developers choose not to create the first cell separately from the others because "it's too much work" to have one `<th>` and everything else be a `<td>`. Really happy to see you taking the time to do this (especially since it's actually pretty straightforward).

BatemanVO reviewed (commented):
Great work on the refactor! There are a lot of complex moving parts in this ticket and you've done a good job creating an end product that looks to meet business needs.

You're also making good progress on keeping your logic to named methods and I think your naming conventions are good.

aggiewb committed:
Convert calculateSimulatedCookiesPerHour() into a prototype

method
Change all references of location to this as appropriate
Rearrange all prototype method to be placed together

aggiewb committed:
Change openHours to a global contstant OPEN_HOURS

Update setting a value to an index i to using .push() for clarity

aggiewb committed:
Remove js assignment of caption content, and add directly to

html caption element

aggiewb committed:
Move the appending of the tableRow to the end of the func

so it doesn't trigger a reflow each time a newHourHead is appended

aggiewb committed:
Cache frequently used :00pm and :00am in strings

aggiewb committed:
Move row scope from the tr to the th

aggiewb committed:
Move row scope from the tr to the th

aggiewb committed:
Update for loop conditional to be set to the OPEN_HOURS

constant to make it clearer how many times the for loop will run

aggiewb committed:
Create function createDataCell() and replace all instances of td

being created with the function
Update createStoreTotals() to calculateStoreTotal() as a prototype method and append the cell to the row in createStoreTableContents()

aggiewb committed:
Remove totalOfAllStoresDaily() function and use

createDataCell() inside of createStoreTableContents to append that cell to the table

aggiewb committed:
Update createDataCell() to accept an optional parameter of

table scope type
Update all instance of table cell creation with a call of createDataCell()
Remove createDailyLocationTotalHeader() as createDataCell() does all the intended "work"
Remove all caches of createDataCell() and use directly in the append to the row

aggiewb committed:
Add conditional to tfoot that adds a class to highlight the last

row appropriately if more stores are added or removed

BatemanVO reviewed (approved):
Great work on the changes - looks like we cut about 10% of our code through refactoring!

Go ahead and move this ticket to QA/UAT and assign me once you've merged.
